streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java

```diff
@@ -230,18 +230,25 @@ public UUID initializeProcessId() {
     public File getOrCreateDirectoryForTask(final TaskId taskId) {
         final File taskParentDir = getTaskDirectoryParentName(taskId);
         final File taskDir = new File(taskParentDir, StateManagerUtil.toTaskDirString(taskId));
-        if (hasPersistentStores && !taskDir.exists()) {
-            synchronized (taskDirCreationLock) {
-                // to avoid a race condition, we need to check again if the directory does not exist:
-                // otherwise, two threads might pass the outer `if` (and enter the `then` block),
-                // one blocks on `synchronized` while the other creates the directory,
-                // and the blocking one fails when trying to create it after it's unblocked
-                if (!taskParentDir.exists() && !taskParentDir.mkdir()) {
-                    throw new ProcessorStateException(
+        if (hasPersistentStores) {
+            if (!taskDir.exists()) {
+                synchronized (taskDirCreationLock) {
+                    // to avoid a race condition, we need to check again if the directory does not exist:
+                    // otherwise, two threads might pass the outer `if` (and enter the `then` block),
+                    // one blocks on `synchronized` while the other creates the directory,
+                    // and the blocking one fails when trying to create it after it's unblocked
+                    if (!taskParentDir.exists() && !taskParentDir.mkdir()) {
+                        throw new ProcessorStateException(
                             String.format("Parent [%s] of task directory [%s] doesn't exist and couldn't be created",
-                                    taskParentDir.getPath(), taskDir.getPath()));
+                                taskParentDir.getPath(), taskDir.getPath()));
+                    }
+                    if (!taskDir.exists() && !taskDir.mkdir()) {
+                        throw new ProcessorStateException(
+                            String.format("task directory [%s] doesn't exist and couldn't be created", taskDir.getPath()));
+                    }
                 }
-                if (!taskDir.exists() && !taskDir.mkdir()) {
+            } else {
+                if (!taskDir.isDirectory()) {
```

Comment:
Same here, this exception message does not apply to the case this is trying to catch

Reply:
+1

Reply:
nit: we could use `else if` here.